File: a2/java/quicksort/Sort.java
/*
 * Alexander Marshall - 100487187
 * Compile Errors: <5
 * Runtime Errors: 
 */

package quicksort;
import java.util.Random;

public class Sort<T extends Comparable<T>> {
    private T[] input;
    private boolean ascend;
    Random random;

    public Sort(T[] input, boolean ascend) {
        this.input = input;
        this.ascend = ascend;
        random = new Random();
    }

    //boolean ascend determines sort order
    public int compare(T a, T b) {
        if(ascend) {
            return a.compareTo(b);
        } else {
            return -(a.compareTo(b));
        }
    }

    //Prints arrays of length <= 30
    private void print() {
        if(input.length <= 30) {
            for(int i = 0; i < input.length; i++) {
                System.out.print(input[i] + " ");
            }
            System.out.println("");
        }
    }

    public T[] sort() {
        this.print();
        System.out.print("Sorting... ");
        quicksort(0, input.length-1);
        System.out.println("Done! ");
        this.print();
        return input;
    }

    private void quicksort(int start, int end) {
        if(start < end) {
            T pivot = input[end];
            int i = start; int j = end;

            while(i != j) {
                if(compare(input[i], pivot) < 0) {
                    i++;
                } else {
                    input[j] = input[i];
                    input[i] = input[j-1];
                    j--;
                }
            }
            input[j] = pivot;
            quicksort(start, j-1);
            quicksort(j+1, end);
        }
    }
}

File: a2/scala/quicksort/Sort.scala
//Alexander Marshall 100487187
//Compile Errors: <5
//Runtime Errors:

package quicksort {
    class Quick[T](input: Array[T])(implicit o: Ordering[T]) {
        def go() = {
            val s = new Sort(input)
            s.sort()
        }
    }

    class Sort[T <% Ordered[T]](input: Array[T])(implicit o: Ordering[T]) {
        def swap(a:Int, b:Int) = {
            var temp:T = input(a)
            input(a) = input(b)
            input(b) = temp
        }

        def printArr() = {
            if(input.length < 30)
                println(input.mkString(" "))
        }
        
        def sort () = {
            this.printArr()
            print("Sorting... ")
            quicksort(0, input.length-1)
            println("Done!")
            this.printArr()
        }

        def quicksort(start:Int, end:Int) {
            if(start < end) {
                val pivot:T = input(end)
                var i:Int = start
                var j:Int = end

                while(i != j) {
                    if(input(i) < pivot) {
                        i = i+1
                    } else {
                        input(j) = input(i)
                        input(i) = input(j-1)
                        j = j-1
                    }
                }

                input(j) = pivot
                quicksort(start, j-1)
                quicksort(j+1, end)
            }
        }
    }
}
